docs/proposals/diagnose-experience.md

```diff
@@ -0,0 +1,247 @@
+
+## Guiding Principles
+
+Here are the guiding principles to print out debug logs.
+
+### 1. **Timestamp Each Log Entry**
+- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.
+  - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`
+
+### 2. **Capture API-Specific Details**
+- **API Requests:** Log detailed information about API requests made to the registry server, including the HTTP method, endpoint, headers, and body (excluding sensitive information).
+  - Example: `DEBUG: [HTTPRequest] POST /v2/oras-demo/blobs/uploads/ Headers: {Content-Length: 524288}`
+  
+- **API Responses:** Log details about the API responses received, including status codes, headers, and response body (excluding sensitive information).
+  - Example: `DEBUG: [HTTPResponse] Status: 201 Created, Headers: {Location: /v2/oras-demo/blobs/uploads/uuid}, Body: {}`
+
+### 3. **Log Before and After Critical Operations**
+- **Operation Logs:** Log before performing critical operations and after completing them, including success or failure status.
+  - Example: `DEBUG: Starting upload of layer 2 of 3 for repository oras-demo`
+  - Example: `DEBUG: Successfully uploaded layer 2 of 3 for repository oras-demo`
+
+- **State Logs:** Log important state information before and after key operations or decisions.
+  - Example: `DEBUG: Current retry attempt: 1, Max retries: 3`
+
+### 4. **Error and Exception Handling**
+- **Catch and Log Exceptions:** Always catch exceptions and log them with relevant context and stack traces.
+  - Example: `ERROR: Exception occurred in fetchManifest: Network timeout while accessing /v2/oras-demo/manifests/latest`
+  
+- **Error Codes:** Include specific error codes to facilitate quick identification and resolution.
+  - Example: `ERROR: [ErrorCode: 504] Network timeout while accessing /v2/oras-demo/manifests/latest`
+
+### 5. **Environment and Configuration Details**
+- **Environment Information:** Log details about the environment where the tool is running, such as OS version and architecture, tool version, and environment variables (excluding sensitive data).
+  - Example: `DEBUG: Running on OS: Ubuntu 20.04, Arch: Linux AMD64 ORAS version: v1.2.0`
+
+- **Configuration Settings:** Include current configuration settings that might affect the execution or behavior of the tool.
+  - Example: `DEBUG: Configuration - Registry URL: https://myregistry.io, Timeout: 30s, RetryCount: 3`
+
+### 5. **Include Actionable Information**
+- **Guidance:** Where possible, provide suggestions within logs about potential fixes or next steps for resolving issues.
+  - Example: `DEBUG: [Action] Check network connectivity and retry the operation. Error occurred while accessing /v2/oras-demo/manifests/latest`
+
+- **Diagnostic Tips:** Include information that can assist in diagnosing issues like configuration settings, environment variables, or system states.
```

Comment:
What is the difference between this and **5. Environment and Configuration Details**?

Reply:
I removed this point as it is duplicated.